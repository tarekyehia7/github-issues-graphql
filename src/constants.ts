export const constants = {
	repository: 'facebook',
	projectName: 'react',
	searchTerm: 'title,body',
	issuesPerPage: 11,
	githubUrl: 'https://github.com',
};
