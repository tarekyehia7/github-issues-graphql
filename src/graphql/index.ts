import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
	uri: process.env.REACT_APP_API_URL,
});

const authLink = setContext((_, { headers }) => {
	const token = process.env.REACT_APP_GITHUB_AUTH_TOKEN;

	return {
		headers: {
			...headers,
			authorization: token ? `Bearer ${token}` : '',
		},
	};
});

export const apolloClient = new ApolloClient({
	link: authLink.concat(httpLink),
	cache: new InMemoryCache(),
	connectToDevTools: true,
});
