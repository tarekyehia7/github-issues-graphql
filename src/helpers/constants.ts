export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};