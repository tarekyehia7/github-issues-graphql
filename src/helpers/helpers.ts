import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
