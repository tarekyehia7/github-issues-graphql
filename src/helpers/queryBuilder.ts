import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};