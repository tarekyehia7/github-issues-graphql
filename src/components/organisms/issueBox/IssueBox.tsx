import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';
import React from 'react';

import {
	Box,
	CommentsIconStyled,
	Description,
	Header,
	OpenIssueIconStyled,
	SkipIconStyled,
	CommentsSection,
} from './IssueBox.styled';
import { Link, LinkType } from '../../atoms/link/Link';
import { IssueState } from '../../../graphql/generatedTypes/graphql';
import { constants } from '../../../constants';

type IssueBoxProps = {
	state: IssueState;
	title: string;
	totalCount: number;
	issueId: number;
	createdAt: string;
	authorName: string;
	authorUrl: string;
};

export const IssueBox = ({
	state,
	title,
	totalCount,
	issueId,
	createdAt,
	authorName,
	authorUrl,
}: IssueBoxProps) => {
	return (
		<Box>
			<Header>
				<Link
					type={LinkType.Title}
					to={`/${constants.repository}/${constants.projectName}/issue/${issueId}`}
				>
					{state === IssueState.Open && <OpenIssueIconStyled />}
					{state === IssueState.Closed && <SkipIconStyled />}
					{title}
				</Link>
				{totalCount > 0 && (
					<CommentsSection>
						<Link
							type={LinkType.Normal}
							to={`/${constants.repository}/${constants.projectName}/issue/${issueId}`}
						>
							<CommentsIconStyled />
							<span>{totalCount}</span>
						</Link>
					</CommentsSection>
				)}
			</Header>
			<Description>
				<span>#{issueId} Opened </span>
				{formatDistanceToNowStrict(new Date(createdAt))}
				<span> ago by </span>
				<Link type={LinkType.Normal} to={authorUrl}>
					{authorName}
				</Link>
			</Description>
		</Box>
	);
};
