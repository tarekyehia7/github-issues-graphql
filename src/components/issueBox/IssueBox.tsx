import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';
import React from 'react';
import { useNavigate } from 'react-router-dom';

import {
	Box,
	Ahref,
	Comments,
	CommentsIconStyled,
	Description,
	Header,
	OpenIssueIconStyled,
	SkipIconStyled,
	Title,
} from './IssueBox.styled';
import { IssueState } from '../../graphql/generatedTypes/graphql';
import { constants } from '../../helpers/constants';

type IssueBoxProps = {
	state: IssueState;
	title: string;
	totalCount: number;
	issueId: number;
	createdAt: string;
	authorName: string;
	authorUrl: string;
};

export const IssueBox = ({
	state,
	title,
	totalCount,
	issueId,
	createdAt,
	authorName,
	authorUrl,
}: IssueBoxProps) => {
	const navigate = useNavigate();

	const navigateToIssue = () => {
		navigate(`/${constants.repository}/${constants.projectName}/issue/${issueId}`);
	};

	return (
		<Box>
			<Header>
				<Title onClick={navigateToIssue}>
					{state === IssueState.Open && <OpenIssueIconStyled />}
					{state === IssueState.Closed && <SkipIconStyled />}
					{title}
				</Title>
				{totalCount > 0 && (
					<Comments onClick={navigateToIssue}>
						<CommentsIconStyled />
						{totalCount}
					</Comments>
				)}
			</Header>
			<Description>
				#{issueId} Opened
				{formatDistanceToNowStrict(new Date(createdAt))}
				ago by
				<Ahref href={authorUrl}>{authorName}</Ahref>
			</Description>
		</Box>
	);
};
