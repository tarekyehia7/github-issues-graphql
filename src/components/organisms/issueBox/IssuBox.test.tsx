import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';

import { IssueBox, IssueBoxProps } from './IssueBox';
import { PageWithTheme } from '../../../helpers/testing/helpers';
import { IssueState } from '../../../graphql/generatedTypes/graphql';
import { formatDate } from '../../../helpers/helpers';
import { constants } from '../../../constants';

const props: IssueBoxProps = {
	state: IssueState.Open,
	title: 'Bug: <img/> renders twice in firefox with react 18',
	totalCount: 10,
	issueId: 25276,
	createdAt: '2022-09-09T13:25:28Z',
	authorName: 'tyehia',
	authorUrl: 'https://github.com/tarekyehia7',
};
const IssueBoxWithTheme = () => {
	return (
		<PageWithTheme>
			<IssueBox {...props} />
		</PageWithTheme>
	);
};

describe('<IssueBox />', () => {
	it('Should match snapshot', async () => {
		// const domTree = await renderer.create(<IssueBoxWithTheme />).toJSON();
        const { container } = render(<IssueBoxWithTheme />);
		expect(container).toMatchSnapshot();
	});

	it('Should have correct Title', async () => {
		const { getByText } = render(<IssueBoxWithTheme />);

		expect(getByText(props.title)).toBeInTheDocument();
	});

	it('Should have correct description', async () => {
		const { getByText } = render(<IssueBoxWithTheme />);

		expect(getByText(`#${props.issueId} Opened`)).toBeInTheDocument();
		expect(getByText(formatDate(props.createdAt))).toBeInTheDocument();
	});

	it('Should have comments section', async () => {
		const { repository, projectName } = constants;
		const { getByTestId } = render(<IssueBoxWithTheme />);

		expect(getByTestId('comments-section')).toBeInTheDocument();
		expect(getByTestId('comments-section').firstChild?.lastChild).toHaveTextContent(
			`${props.totalCount}`,
		);
		expect(getByTestId('comments-section').firstChild).toHaveAttribute(
			'href',
			`/${repository}/${projectName}/issue/${props.issueId}`,
		);
	});
});
