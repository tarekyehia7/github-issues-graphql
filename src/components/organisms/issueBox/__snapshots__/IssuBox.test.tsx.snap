// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IssueBox /> Should match snapshot 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16px;
  font-weight: 600;
  color: #24292f;
  margin-top: 0;
  cursor: pointer;
  background: transparent;
  margin: 1rem;
  margin-top: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover {
  color: #0969da;
}

.c9 {
  color: #57606a;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c9:hover {
  color: #0969da;
}

.c0 {
  padding: 16px;
  list-style-type: none;
  border-top-color: #d8dee4;
  border-top-style: solid;
  border-top-width: 1px;
}

.c0:hover {
  background-color: #f6f8fa;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  margin-right: 0.2rem;
  height: 1rem;
  width: 1rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  cursor: pointer;
}

.c4:hover > a {
  color: #0969da;
}

.c4:hover .c6 {
  fill: #0969da;
}

.c8 {
  color: #57606a;
  margin-left: 1.7rem;
  font-size: 13px;
}

.c3 {
  min-height: 1rem;
  min-width: 2rem;
  margin-top: 0.07rem;
  fill: #cf222e;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0"
    data-testid="issue-box-25276"
  >
    <div
      class="c1"
    >
      <a
        class="c2"
        href="/facebook/react/issue/25276"
      >
        <svg
          aria-hidden="true"
          class="c3"
          data-view-component="true"
          height="16"
          version="1.1"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
          />
          <path
            d="M8 0a8 8 0 100 16A8 8 0 008 0zM1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0z"
          />
        </svg>
        Bug: &lt;img/&gt; renders twice in firefox with react 18
      </a>
      <div
        class="c4"
        data-testid="comments-section"
      >
        <a
          class="c2 c5"
          href="/facebook/react/issue/25276"
        >
          <svg
            aria-hidden="true"
            class="c6 c7"
            data-view-component="true"
            height="16"
            version="1.1"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M2.75 2.5a.25.25 0 00-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 01.75.75v2.19l2.72-2.72a.75.75 0 01.53-.22h4.5a.25.25 0 00.25-.25v-7.5a.25.25 0 00-.25-.25H2.75zM1 2.75C1 1.784 1.784 1 2.75 1h10.5c.966 0 1.75.784 1.75 1.75v7.5A1.75 1.75 0 0113.25 12H9.06l-2.573 2.573A1.457 1.457 0 014 13.543V12H2.75A1.75 1.75 0 011 10.25v-7.5z"
            />
          </svg>
          <span>
            10
          </span>
        </a>
      </div>
    </div>
    <div
      class="c8"
    >
      <span>
        #
        25276
         Opened 
      </span>
      8 days
      <span>
         ago by 
      </span>
      <a
        class="c9"
        href="https://github.com/tarekyehia7"
      >
        tyehia
      </a>
    </div>
  </div>
</div>
`;
